### PR TITLE
bpf: host: also handle from-egress proxy traffic

### PR DESCRIPTION
**Pull request by jschwinger233**
See commit messages for details.

```release-note
Fixes proxy issues in egress direction
```

### CODE REVIEW

**Review by julianwiedmann — APPROVED** [review]

**Comment by jrajahalme** [comment]
@jschwinger233 I wonder if the socket mark condition here could be changed that it only applies to the forward path, and not to the return packet path? This change has caused issues with proxy return packets not getting to the proxy, likely due to the encryption mark being left on the packet after the return packets have been decrypted. This has caused issues like this: https://github.com/cilium/cilium/issues/27762#issuecomment-2118311808. Currently fixing these by making sure the CT entry exists and has the `ProxyRedirect` flag set (e.g., https://github.com/cilium/cilium/pull/32614).

So in summary I see two possible ways to remedy this:
1) make sure the encryption marks are cleared after decryption of the return packets (preferable, IMO)
2) modify the conditions on the socket match rule so that they only apply to the forward path packets

**Comment by jschwinger233** [comment]
@jrajahalme Sorry but I'm not sure how this PR causes https://github.com/cilium/cilium/issues/27762 which doesn't enable IPsec. Most changes this PR made is under the guard of `#if defined(ENABLE_IPSEC)`, expect for `magic = MARK_MAGIC_PROXY_EGRESS`, which only makes a difference for local_delivery. 

Do you think it's possible that https://github.com/cilium/cilium/pull/32367 has something to do with it?
